pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -1283,7 +1288,7 @@ private CompletableFuture<Optional<Topic>> createNonPersistentTopic(String topic
                     }).exceptionally(ex -> {
                 log.warn("Replication check failed. Removing topic from topics list {}, {}", topic, ex.getCause());
                 nonPersistentTopic.stopReplProducers().whenComplete((v, exception) -> {
-                    pulsar.getExecutor().execute(() -> topics.remove(topic, topicFuture));
+                    pulsar.getExecutor().execute(() -> topics.remove(topic));
```

Comment:
In the past there have been races which has been resolved by removing a specific value. We should ensure that this change doesn't cause regressions.

Reply:
> In the past there have been races which has been resolved by removing a specific value. We should ensure that this change doesn't cause regressions.

I don't think there should be any regressions now because previously, calling 'remove' at the start of 'getTopic' led to a race condition.

https://github.com/apache/pulsar/blob/400a286ca09d4f45f388616fd996f0605f19e5a4/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L1011-L1012

Do you remember the previous PR? I'll see if there's a unit test to cover it and I can verify it.

Reply:
Please check the commit that added that line.

Reply:
I have not found any reason for it.

This approach has been the same from the beginning, tracing back to the `initial import`. It seems that everyone just continued with the same style to maintain consistency.

https://github.com/apache/pulsar/pull/538/files#diff-0210356c8a88e4efa89eb769a027fa6c166db479dbad8bbbbc704c6ed6e317f5R463


However, in reality, if you want to compare topic futures, it is easy to make mistakes in deletion. For examples:

``` java

    static CompletableFuture<Optional<Integer>> m1 = CompletableFuture.completedFuture(Optional.of(100));
    static CompletableFuture<Optional<String>> m2 = CompletableFuture.completedFuture(Optional.of("value1"));

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        ConcurrentOpenHashMap<String, CompletableFuture<Optional<String>>> maps = 
                ConcurrentOpenHashMap.<String, CompletableFuture<Optional<String>>>newBuilder()
                .build();
        maps.computeIfAbsent("test1", (key) -> {
            return getM1().thenCompose(__ -> {
                return getM2(); 
            });
        });
        System.out.println("debug m1" + m1 + "@" + System.identityHashCode(m1));
        System.out.println("debug m2" + m2 + "@" + System.identityHashCode(m2));
        
        CompletableFuture<Optional<String>> test1 = maps.get("test1");
        System.out.println("debug1: " + test1.get().get());
        System.out.println("debug test1:" + test1 + "@" + System.identityHashCode(test1));

        maps.remove("test1", m2);
        CompletableFuture<Optional<String>> test2 = maps.get("test1");
        System.out.println("debug2: " + test2.get().get());
        System.out.println("debug test2:" + test2 + "@" + System.identityHashCode(test2));
    }

    public static CompletableFuture<Optional<Integer>> getM1() {

        return m1;
    }
    
    public static CompletableFuture<Optional<String>> getM2() {
        return m2;
    }
```

A similar code was written here. So, what's in the cache is not the original `topicFuture`, its a transformed future.
https://github.com/apache/pulsar/blob/21596227415968dd9b9219779657ff74decfa45b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L1042-L1073

Reply:
> I have not found any reason for it.

There are a few issues that @poorbarcode has fixed by removing the specific value instead of removing by key.
Perhaps he could share more details.

> However, in reality, if you want to compare topic futures, it is easy to make mistakes in deletion.

That's true. The compiler won't catch the mistakes, but we should have tests to cover that.

Reply:
> There are a few issues that @poorbarcode has fixed by removing the specific value instead of removing by key.
Perhaps he could share more details.

Thanks reply, I guess it is related to this PR: https://github.com/apache/pulsar/pull/17524

But this PR doesn't modify this behavior.